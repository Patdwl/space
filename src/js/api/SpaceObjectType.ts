export enum SpaceObjectType {
  UNKNOWN = 0,
  PAYLOAD = 1,
  ROCKET_BODY = 2,
  DEBRIS = 3,
  SPECIAL = 4,
  RADAR_MEASUREMENT = 5,
  RADAR_TRACK = 6,
  RADAR_OBJECT = 7,
  BALLISTIC_MISSILE = 8,
  STAR = 9,
  INTERGOVERNMENTAL_ORGANIZATION = 10,
  SUBORBITAL_PAYLOAD_OPERATOR = 11,
  PAYLOAD_OWNER = 12,
  METEOROLOGICAL_ROCKET_LAUNCH_AGENCY_OR_MANUFACTURER = 13,
  PAYLOAD_MANUFACTURER = 14,
  LAUNCH_AGENCY = 15,
  LAUNCH_SITE = 16,
  LAUNCH_POSITION = 17,
  LAUNCH_FACILITY = 18,
  CONTROL_FACILITY = 19,
  GROUND_SENSOR_STATION = 20,
  OPTICAL = 21,
  MECHANICAL = 22,
  PHASED_ARRAY_RADAR = 23,
  OBSERVER = 24,
  BISTATIC_RADIO_TELESCOPE = 25,
  COUNTRY = 26,
  LAUNCH_VEHICLE_MANUFACTURER = 27,
  ENGINE_MANUFACTURER = 28,
}
