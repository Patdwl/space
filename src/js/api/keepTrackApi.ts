import { SatMath } from '../satMath/satMath';
import { KeepTrackApi, SatObject, SettingsManager } from './keepTrackTypes';
import { html } from './templateLiterals';

declare global {
  // eslint-disable-next-line no-unused-vars
  interface Window {
    settingsManager: SettingsManager;
    jQuery: unknown;
    $: unknown;
    gremlins: any;
    randomizer: any;
    keepTrackApi: KeepTrackApi;
    dataLayer: any; // For Google Tag Manager
    _numeric: any;
    satellite: SatMath;
    M: any;
  }
}

export const register = (params: { method: string; cbName: string; cb: any }) => {
  // If this is a valid callback
  if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
    // Add the callback
    keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
  } else {
    throw new Error(`Invalid callback "${params.method}"!`);
  }
};
export const unregister = (params: { method: string; cbName: string }) => {
  // If this is a valid callback
  if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
    for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
      if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
        keepTrackApi.callbacks[params.method].splice(i, 1);
        return;
      }
    }
    // If we got this far, it means we couldn't find the callback
    throw new Error(`Callback "${params.cbName} not found"!`);
  } else {
    // Couldn't find the method
    throw new Error(`Invalid callback "${params.method}"!`);
  }
};
export const isThisJest = () => typeof process !== 'undefined';

export const keepTrackApi: KeepTrackApi = {
  html: html,
  register: register,
  unregister: unregister,
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
    onCruncherReady: [],
    onCruncherMessage: [],
    uiManagerInit: [],
    uiManagerOnReady: [],
    bottomMenuClick: [],
    hideSideMenus: [],
    nightToggle: [],
    orbitManagerInit: [],
    adviceReady: [],
    drawManagerLoadScene: [],
    drawOptionalScenery: [],
    updateLoop: [],
    rmbMenuActions: [],
    rightBtnMenuAdd: [],
    updateDateTime: [],
    uiManagerFinal: [],
    loadCatalog: [],
    resetSensor: [],
    setSensor: [],
    onDrawLoopComplete: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      keepTrackApi.programs.soundManager.play('whoosh');
      keepTrackApi.callbacks.selectSatData.forEach((cb: any) => cb.cb(sat, satId));
    },
    updateSelectBox: (sat: any) => {
      keepTrackApi.callbacks.updateSelectBox.forEach((cb: any) => cb.cb(sat));
    },
    onCruncherReady: () => {
      keepTrackApi.callbacks.onCruncherReady.forEach((cb: any) => cb.cb());
    },
    onCruncherMessage: () => {
      keepTrackApi.callbacks.onCruncherMessage.forEach((cb: any) => cb.cb());
    },
    uiManagerInit: () => {
      keepTrackApi.callbacks.uiManagerInit.forEach((cb: any) => cb.cb());
    },
    uiManagerOnReady: () => {
      keepTrackApi.callbacks.uiManagerOnReady.forEach((cb: any) => cb.cb());
    },
    bottomMenuClick: (iconName: string) => {
      keepTrackApi.programs.soundManager.play('genericBeep');
      keepTrackApi.callbacks.bottomMenuClick.forEach((cb: any) => cb.cb(iconName));
    },
    hideSideMenus: () => {
      keepTrackApi.callbacks.hideSideMenus.forEach((cb: any) => cb.cb());
    },
    nightToggle: (gl: any, nightTexture: any, texture: any) => {
      keepTrackApi.callbacks.nightToggle.forEach((cb: any) => cb.cb(gl, nightTexture, texture));
    },
    orbitManagerInit: () => {
      keepTrackApi.callbacks.orbitManagerInit.forEach((cb: any) => cb.cb());
    },
    adviceReady: () => {
      keepTrackApi.callbacks.adviceReady.forEach((cb: any) => cb.cb());
    },
    drawManagerLoadScene: () => {
      keepTrackApi.callbacks.drawManagerLoadScene.forEach((cb: any) => cb.cb());
    },
    drawOptionalScenery: () => {
      keepTrackApi.callbacks.drawOptionalScenery.forEach((cb: any) => cb.cb());
    },
    updateLoop: () => {
      keepTrackApi.callbacks.updateLoop.forEach((cb: any) => cb.cb());
    },
    rmbMenuActions: (menuName: string, satnum = -1) => {
      keepTrackApi.callbacks.rmbMenuActions.forEach((cb: any) => cb.cb(menuName, satnum));
    },
    rightBtnMenuAdd: () => {
      keepTrackApi.callbacks.rightBtnMenuAdd.forEach((cb: any) => cb.cb());
    },
    updateDateTime: (date: Date) => {
      keepTrackApi.callbacks.updateDateTime.forEach((cb: any) => cb.cb(date));
    },
    uiManagerFinal: () => {
      keepTrackApi.callbacks.uiManagerFinal.forEach((cb: any) => cb.cb());
    },
    loadCatalog: async (): Promise<SatObject[]> => keepTrackApi.callbacks.loadCatalog[0].cb(),
    resetSensor: () => {
      keepTrackApi.callbacks.resetSensor.forEach((cb: any) => cb.cb());
    },
    setSensor: (sensor: any, id: number) => {
      keepTrackApi.callbacks.setSensor.forEach((cb: any) => cb.cb(sensor, id));
    },
    onDrawLoopComplete: () => {
      keepTrackApi.callbacks.onDrawLoopComplete.forEach((cb: any) => cb.cb());
    },
  },
  programs: <any>{},
};

// First time we call this module we should make it available to the rest of the application
if (!window.keepTrackApi) window.keepTrackApi = keepTrackApi;
