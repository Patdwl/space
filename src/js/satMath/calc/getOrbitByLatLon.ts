import { SatObject } from '@app/js/api/keepTrackTypes';
import { RAD2DEG } from '@app/js/lib/constants';
import { stringPad } from '@app/js/lib/helpers';
import { EciVec3, SatelliteRecord } from 'ootk';
import { satellite } from '../satMath';
import { calculateTimeVariables } from './calculateTimeVariables';

enum PropagationOptions {
  MeanAnomaly = 1,
  RightAscensionOfAscendingNode = 2,
  ArgumentOfPerigee = 3,
}

enum PropagationResults {
  Near = 0,
  Success = 1,
  Error = 2,
  Far = 3,
}

// prettier-ignore
export const getOrbitByLatLon = (
  sat: SatObject,
  goalLat: number,
  goalLon: number,
  goalDirection: 'N' | 'S',
  now: Date,
  goalAlt?: number,
  raanOffset?: number
): [string, string] => { // NOSONAR
  let newMeana: string = null;
  let newArgPer: string = null;
  goalAlt = goalAlt || null;
  raanOffset = raanOffset || 0;

  let lastLat: number = null;
  let currentDirection: 'N' | 'S';

  // Parse some values used in creating new TLEs
  const intl = sat.TLE1.substr(9, 8);
  const epochyr = sat.TLE1.substr(18, 2);
  const epochday = sat.TLE1.substr(20, 12);
  const meanmo = sat.TLE2.substr(52, 11);
  const inc = stringPad.pad0((sat.inclination * RAD2DEG).toFixed(4), 8);
  const ecen = sat.eccentricity.toFixed(7).substr(2, 7);
  // Disregarding the first and second derivatives of mean motion
  // Just keep whatever was in the original TLE
  const TLE1Ending = sat.TLE1.substr(32, 39);

  // //////////////////////////////////////////////////////////////////////////////
  // //////////////////////////////////////////////////////////////////////////////
  // These four functions get called during the bottom loop

  /**
   * Rotating the mean anomaly adjusts the latitude (and longitude) of the satellite.
   * @param {number} meana - This is the mean anomaly (where it is along the orbital plane)
   * @returns {PropagationResults} This number tells the main loop what to do next
   */
  const meanaCalc = (meana: number) => {
    let satrec = sat.satrec || <SatelliteRecord>satellite.twoline2satrec(sat.TLE1, sat.TLE2); // perform and store sat init calcs

    meana = meana / 10;
    const meanaStr = stringPad.pad0(meana.toFixed(4), 8);

    const raan = stringPad.pad0((sat.raan * RAD2DEG).toFixed(4), 8);

    const argPe = newArgPer ? stringPad.pad0((parseFloat(newArgPer) / 10).toFixed(4), 8) : stringPad.pad0((sat.argPe * RAD2DEG).toFixed(4), 8);

    const _TLE1Ending = sat.TLE1.substr(32, 39);
    const TLE1 = '1 ' + sat.sccNum + 'U ' + intl + ' ' + epochyr + epochday + _TLE1Ending; // M' and M'' are both set to 0 to put the object in a perfect stable orbit
    const TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + raan + ' ' + ecen + ' ' + argPe + ' ' + meanaStr + ' ' + meanmo + '    10';

    satrec = satellite.twoline2satrec(TLE1, TLE2);
    const results = getOrbitByLatLonPropagate(now, satrec, PropagationOptions.MeanAnomaly);
    if (results === PropagationResults.Success) {
      sat.TLE1 = TLE1;
      sat.TLE2 = TLE2;
      newMeana = meanaStr;
    }
    return results;
  };

  /**
   * We need to adjust the argument of perigee to align a HEO orbit with the desired launch location
   * @param {string} argPe - This is the guess for the argument of perigee (where the lowest part of the orbital plane is)
   * @returns {PropagationResults} This number tells the main loop what to do next
   */
  const argPerCalc = (argPe: string): PropagationResults => {
    const meana = newMeana;
    const raan = stringPad.pad0((sat.raan * RAD2DEG).toFixed(4), 8);
    argPe = stringPad.pad0((parseFloat(argPe) / 10).toFixed(4), 8);

    // Create the new TLEs
    const TLE1 = '1 ' + sat.sccNum + 'U ' + intl + ' ' + epochyr + epochday + TLE1Ending;
    const TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + raan + ' ' + ecen + ' ' + argPe + ' ' + meana + ' ' + meanmo + '    10';

    // Calculate the orbit
    const satrec = <SatelliteRecord>satellite.twoline2satrec(TLE1, TLE2);

    // Check the orbit
    const results = getOrbitByLatLonPropagate(now, satrec, PropagationOptions.ArgumentOfPerigee);
    if (results === PropagationResults.Success) {
      sat.TLE1 = TLE1;
      sat.TLE2 = TLE2;
      newArgPer = argPe;
    }
    return results;
  };

  /**
   * Rotating the mean anomaly adjusts the latitude (and longitude) of the satellite.
   * @param {number} raan - This is the right ascension of the ascending node (where it rises above the equator relative to a specific star)
   * @param {number} raanOffsetIn - This allows the main thread to send a guess of the raan
   * @returns {PropagationResults} This number tells the main loop what to do next
   */
  const raanCalc = (raan: number, raanOffsetIn: number): PropagationResults => {
    const origRaan = raan;
    raan = raan / 100;
    raan = raan > 360 ? raan - 360 : raan;

    const raanStr = stringPad.pad0(raan.toFixed(4), 8);

    // If we adjusted argPe use the new one - otherwise use the old one
    const argPe = newArgPer ? stringPad.pad0((parseFloat(newArgPer) / 10).toFixed(4), 8) : stringPad.pad0((sat.argPe * RAD2DEG).toFixed(4), 8);

    const TLE1 = '1 ' + sat.sccNum + 'U ' + intl + ' ' + epochyr + epochday + TLE1Ending; // M' and M'' are both set to 0 to put the object in a perfect stable orbit
    const TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + raanStr + ' ' + ecen + ' ' + argPe + ' ' + newMeana + ' ' + meanmo + '    10';

    const satrec = <SatelliteRecord>satellite.twoline2satrec(TLE1, TLE2);
    const results = getOrbitByLatLonPropagate(now, satrec, PropagationOptions.RightAscensionOfAscendingNode);

    // If we have a good guess of the raan, we can use it, but need to apply the offset to the original raan
    if (results === PropagationResults.Success) {
      raan = origRaan / 100 + raanOffsetIn;
      raan = raan > 360 ? raan - 360 : raan;
      raan = raan < 0 ? raan + 360 : raan;

      const _raanStr = stringPad.pad0(raan.toFixed(4), 8);

      const _TLE2 = '2 ' + sat.sccNum + ' ' + inc + ' ' + _raanStr + ' ' + ecen + ' ' + argPe + ' ' + newMeana + ' ' + meanmo + '    10';

      sat.TLE1 = TLE1;
      sat.TLE2 = _TLE2;
    }
    return results;
  };

  const getOrbitByLatLonPropagate = (nowIn: Date, satrec: SatelliteRecord, type: PropagationOptions): PropagationResults => {
    const { m, gmst } = calculateTimeVariables(nowIn, satrec);
    const positionEci = <EciVec3>satellite.sgp4(satrec, m).position;
    if (isNaN(positionEci.x) || isNaN(positionEci.y) || isNaN(positionEci.z)) {
      return PropagationResults.Error;
    }
    const gpos = satellite.eciToGeodetic(positionEci, gmst);

    let { lat, lon, alt } = gpos;
    lat = satellite.degreesLat(lat);
    lon = satellite.degreesLong(lon);

    // Set it the first time
    lastLat = lastLat ? lastLat : lat;

    if (type === PropagationOptions.MeanAnomaly) {
      if (lat === lastLat) {
        return 0; // Not enough movement, skip this
      }

      if (lat > lastLat) {
        currentDirection = 'N';
      }
      if (lat < lastLat) {
        currentDirection = 'S';
      }

      lastLat = lat;
    }

    if (type === PropagationOptions.MeanAnomaly && lat > goalLat - 0.15 && lat < goalLat + 0.15) {
      return PropagationResults.Success;
    }

    if (type === PropagationOptions.RightAscensionOfAscendingNode && lon > goalLon - 0.15 && lon < goalLon + 0.15) {
      return PropagationResults.Success;
    }

    if (type === PropagationOptions.ArgumentOfPerigee && alt > goalAlt - 30 && alt < goalAlt + 30) {
      return PropagationResults.Success;
    }

    // If current latitude greater than 11 degrees off rotate meanA faster
    if (type === PropagationOptions.MeanAnomaly && !(lat > goalLat - 11 && lat < goalLat + 11)) {
      return PropagationResults.Far;
    }

    // If current longitude greater than 11 degrees off rotate raan faster
    if (type === PropagationOptions.RightAscensionOfAscendingNode && !(lon > goalLon - 11 && lon < goalLon + 11)) {
      return PropagationResults.Far;
    }

    // If current altitude greater than 100 km off rotate augPerigee faster
    if (type === PropagationOptions.ArgumentOfPerigee && (alt < goalAlt - 100 || alt > goalAlt + 100)) {
      return PropagationResults.Far;
    }

    return PropagationResults.Near;
  };
  // //////////////////////////////////////////////////////////////////////////////
  // //////////////////////////////////////////////////////////////////////////////

  let argPerCalcResults: PropagationResults;
  let meanACalcResults: PropagationResults;
  let raanCalcResults: PropagationResults;
  /** Rotate Mean Anomaly 0.1 Degree at a Time for Up To 520 Degrees */
  for (let i = 0; i < 520 * 10; i += 1) {
    meanACalcResults = meanaCalc(i);
    if (meanACalcResults === PropagationResults.Success) {
      if (currentDirection !== goalDirection) {
        // Move 2 Degrees ahead in the orbit to prevent being close on the next lattiude check
        // This happens when the goal latitude is near the poles
        i += 20; // NOSONAR
      } else {
        break; // Stop changing the Mean Anomaly
      }
    }
    if (meanACalcResults === PropagationResults.Far) {
      i += 100; // NOSONAR
    }
  }

  if (meanACalcResults !== PropagationResults.Success) {
    return ['Error', 'Failed to find a solution for Mean Anomaly'];
  }

  // ===== Argument of Perigee Loop =====
  // Applies to eccentric orbits
  if (goalAlt) {
    // Don't Bother Unless Specifically Requested
    meanACalcResults = 0; // Reset meanACalcResults
    for (let i = 0; i < 360 * 10; i += 1) {
      /** Rotate ArgPer 0.1 Degree at a Time for Up To 400 Degrees */
      argPerCalcResults = argPerCalc(i.toString());
      if (argPerCalcResults === PropagationResults.Success) {
        // DEBUG:
        // console.log('Found Correct Alt');
        if (meanACalcResults === PropagationResults.Success) {
          // DEBUG:
          // console.log('Found Correct Lat');

          // DEBUG:
          // console.log('Up Or Down: ' + upOrDown);
          if (currentDirection === goalDirection) {
            // If Object is moving in the goal direction (upOrDown)
            break; // Stop changing ArgPer
          }
        } else {
          // DEBUG:
          // console.log('Found Wrong Lat');
        }
      } else {
        // DEBUG:
        // console.log('Failed Arg of Per Calc');
      }
      if (argPerCalcResults === PropagationResults.Far) {
        // Change ArgPer faster
        i += 5 * 10; // NOSONAR
      }
      if (argPerCalcResults === PropagationResults.Error) {
        return ['Error', 'Failed to find a solution for Argument of Perigee'];
      }

      // ===== Mean Anomaly Loop =====
      for (let j = 0; j < 520 * 10; j += 1) {
        /** Rotate Mean Anomaly 0.1 Degree at a Time for Up To 400 Degrees */
        meanACalcResults = meanaCalc(j);
        if (meanACalcResults === PropagationResults.Success) {
          if (currentDirection !== goalDirection) {
            // If Object is moving opposite of the goal direction (upOrDown)
            // Move 2 Degrees ahead in the orbit to prevent being close on the next lattiude check
            j = j + 20; // NOSONAR
          } else {
            break; // Stop changing the Mean Anomaly
          }
        }
        j = meanACalcResults === PropagationResults.Far ? j + 100 : j; // NOSONAR
        if (meanACalcResults === PropagationResults.Error) {
          return ['Error', ''];
        }
      }
    }

    if (argPerCalcResults !== PropagationResults.Success) {
      return ['Error', 'Failed to find a solution for Argument of Perigee'];
    }
  }

  // ===== Right Ascension Loop =====
  for (let i = 0; i < 520 * 100; i += 1) {
    // 520 degress in 0.01 increments TODO More precise?
    raanCalcResults = raanCalc(i, raanOffset);
    if (raanCalcResults === PropagationResults.Success) {
      break;
    }
    if (raanCalcResults === PropagationResults.Far) {
      i += 10 * 100; // NOSONAR
    }
  }

  if (raanCalcResults !== PropagationResults.Success) {
    return ['Error', 'Failed to find a solution for Right Ascension of Ascending Node'];
  }

  return [sat.TLE1, sat.TLE2];
};
