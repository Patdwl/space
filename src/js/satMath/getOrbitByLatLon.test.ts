import { defaultSat } from '../api/apiMocks';
import * as getOrbitByLatLon from './getOrbitByLatLon';

// @ponicode
describe('getOrbitByLatLon.getOrbitByLatLon', () => {
  test('0', () => {
    let param5: any = new Date('01-01-2020');
    let result: any = getOrbitByLatLon.getOrbitByLatLon(defaultSat, 0.15, -0.85, 'S', param5);
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let param5: any = new Date('01-01-2020');
    let result: any = getOrbitByLatLon.getOrbitByLatLon(defaultSat, 11.0, 11, 'N', param5);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let param5: any = new Date('01-01-2020');
    let result: any = getOrbitByLatLon.getOrbitByLatLon(defaultSat, 0.15, 11.0, 'S', param5);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let param5: any = new Date('01-01-2020');
    let result: any = getOrbitByLatLon.getOrbitByLatLon(defaultSat, 11, 11, 'S', param5, 100, -1.0);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let param5: any = new Date('32-01-2020');
    let result: any = getOrbitByLatLon.getOrbitByLatLon(defaultSat, 11.0, 12, 'S', param5, 100, 0.0);
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    let param5: any = new Date('');
    let result: any = () => getOrbitByLatLon.getOrbitByLatLon(defaultSat, Infinity, Infinity, 'S', param5, Infinity, Infinity);
    expect(result).toMatchSnapshot();
  });
});
