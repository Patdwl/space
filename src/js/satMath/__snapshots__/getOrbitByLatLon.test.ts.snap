// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOrbitByLatLon.getOrbitByLatLon 0 1`] = `
Array [
  "1 00005U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991",
  "2 00005 2958.886 205.7000 0000345 0.000000 17.10000 15.48839820    10",
]
`;

exports[`getOrbitByLatLon.getOrbitByLatLon 1 1`] = `
Array [
  "1 00005U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991",
  "2 00005 2958.886 35.36000 0000345 0.000000 208.6000 15.48839820    10",
]
`;

exports[`getOrbitByLatLon.getOrbitByLatLon 2 1`] = `
Array [
  "1 00005U 98067A   21203.40407588  .00003453  00000-0  71172-4 0  9991",
  "2 00005 2958.886 217.5500 0000345 0.000000 17.10000 15.48839820    10",
]
`;

exports[`getOrbitByLatLon.getOrbitByLatLon 3 1`] = `
Array [
  "Error",
  "",
]
`;

exports[`getOrbitByLatLon.getOrbitByLatLon 4 1`] = `
Array [
  "Error",
  "",
]
`;

exports[`getOrbitByLatLon.getOrbitByLatLon 5 1`] = `[Function]`;
